faunadb-scala/src/main/scala/faunadb/FaunaClient.scala

```diff
@@ -384,38 +385,97 @@ class FaunaClient private (connection: Connection) {
+              case TRANSACTION_ABORTED => TransactionAbortedException(q.description, err.status, q.position);
+              case INVALID_WRITE_TIME =>InvalidWriteTimeException(q.description, err.status, q.position);
+              case INVALID_REF => InvalidReferenceException(q.description, err.status, q.position);
+              case MISSING_IDENTITY => MissingIdentityException(q.description, err.status, q.position);
+              case INVALID_SCOPE => InvalidScopeException(q.description, err.status, q.position);
+              case INVALID_TOKEN => InvalidTokenException(q.description, err.status, q.position);
+              case STACK_OVERFLOW => StackOverflowException(q.description, err.status, q.position);
+              case AUTHENTICATION_FAILED => AuthenticationFailedException(q.description, err.status, q.position);
+              case VALUE_NOT_FOUND => ValueNotFoundException(q.description, err.status, q.position);
+              case INSTANCE_NOT_FOUND => InstanceNotFoundException(q.description, err.status, q.position);
+              case INSTANCE_ALREADY_EXISTS => InstanceAlreadyExistsException(q.description, err.status, q.position);
+              case INSTANCE_NOT_UNIQUE => InstanceNotUniqueException(q.description, err.status, q.position);
+              case INVALID_OBJECT_IN_CONTAINER => InvalidObjectInContainerException(q.description, err.status, q.position);
+              case MOVE_DATABASE_ERROR => MoveDatabaseException(q.description, err.status, q.position);
+              case RECOVERY_FAILED => RecoveryFailedException(q.description, err.status, q.position);
+              case FEATURE_NOT_AVAILABLE => FeatureNotAvailableException(q.description, err.status, q.position);
+              case VALIDATION_FAILED =>
+                val failures: IndexedSeq[String] = q.failures
+                  .map{ v => s"field[${v.field.mkString(",")}] - ${v.code}: ${v.description}"}
+                ValidationFailedException(q.description, err.status, q.position, failures)
+              case _ => UnknownException(err.errors.head.description)
+            }
+        }
+
+        Future.failed(c.getOrElse(UnknownException(err.errors.head.description)))
+      }
+
+    }
+
     statusCode match {
-      case 400 => parseErrorsAndFailWith(new BadRequestException(_))
-      case 401 => parseErrorsAndFailWith(new UnauthorizedException(_))
-      case 403 => parseErrorsAndFailWith(new PermissionDeniedException(_))
-      case 404 => parseErrorsAndFailWith(new NotFoundException(_))
-      case 500 => parseErrorsAndFailWith(new InternalException(_))
-      case 503 => parseErrorsAndFailWith(new UnavailableException(_))
-      case _   => parseErrorsAndFailWith(new UnknownException(_))
+      case 400 => throwQueryError(parseErrors())
+      case 401 => Future.failed(UnauthorizedException("Unauthorized", statusCode))
+      case 403 => throwQueryError(parseErrors())
+      case 404 => throwQueryError(parseErrors())
```

Comment:
We need to add `409` exception here as well (this is a new one), it is already added for java code

Reply:
Done.